File: backend/app/schemas/agent_configs.py
"""智能体配置相关的 Pydantic 模型

用户可配置多种外部智能体 CLI,任务创建时选择一种作为执行器。

安全约定(与 schemas/model_configs.py 一致):
- 响应(GET)绝不回传凭据原文,只返回 has_credentials 布尔,避免凭据经 HTTP 泄露
- 请求(PUT)中 secret 类型字段传空字符串表示"保留已存的值",非空表示"更新为新值"
"""
from pydantic import BaseModel, Field


# ============================================================
# Agent 类型元数据(前端渲染表单用)
# ============================================================


class CredentialField(BaseModel):
    """凭证字段定义(前端据此渲染输入框)"""

    key: str
    label: str
    # secret:敏感凭据(password 输入,加密存储,API 不回传原文)
    # text:非敏感配置(明文存储)
    # select:下拉选择(明文存储,如 provider_type;options 提供可选项)
    type: str
    required: bool = True
    placeholder: str = ""
    help_url: str | None = None
    help_text: str | None = None
    # select 类型专用:可选项列表
    options: list[dict[str, str]] | None = None
    # select 类型专用:默认值(未配置时使用)
    default: str | None = None


class AgentTypeMeta(BaseModel):
    """已注册的 agent 类型元数据(GET /agents/types 返回)"""

    agent_type: str
    display_name: str
    description: str = ""
    credential_fields: list[CredentialField] = Field(default_factory=list)
    help_url: str | None = None


# ============================================================
# 凭证写入请求(PUT /agents/configs/{agent_type})
# ============================================================


class CredentialValue(BaseModel):
    """单个凭证字段的值

    secret 类型:value 传空串表示保留已存值,非空表示更新
    text 类型:value 直接存储(可为空)
    """

    key: str
    value: str = Field(default="", max_length=4096)


class SaveAgentConfigRequest(BaseModel):
    """保存某 agent 配置请求"""

    credentials: list[CredentialValue] = Field(default_factory=list)
    is_active: bool = True


# ============================================================
# 配置读取响应(GET,不含凭据原文)
# ============================================================


class AgentConfigOut(BaseModel):
    """用户已配置的 agent(不含凭据原文)"""

    agent_type: str
    display_name: str
    is_active: bool
    # 是否已填写凭据(任一 secret 字段有值即为 True)
    has_credentials: bool

    model_config = {"from_attributes": True}


class AgentConfigDetailOut(AgentConfigOut):
    """单个 agent 配置详情(含各字段填写状态 + 非 secret 字段回显值)

    安全约定:
    - secret 字段:只返回 credential_status[key]=bool,绝不回传原文
    - text/select 字段:非敏感明文配置,在 credential_values 中回传已配置的值,
      便于前端编辑时回显(用户不用每次重新填 base_url/model/provider_type 等)
    """

    # 各凭证字段的填写状态:{"pat": true, ...}(所有字段)
    credential_status: dict[str, bool] = Field(default_factory=dict)
    # 非 secret(text/select)字段的已配置值:{"base_url": "https://...", "model": "..."}
    # secret 字段不在此 dict 中(绝不回传)
    credential_values: dict[str, str] = Field(default_factory=dict)


class AgentConfigListResponse(BaseModel):
    """用户已配置的 agent 列表"""

    configs: list[AgentConfigOut] = Field(default_factory=list)


# ============================================================
# 凭证测试响应(POST /agents/configs/{agent_type}/test)
# ============================================================


class AgentTestResponse(BaseModel):
    """测试 agent 凭证连通性的响应

    用于「智能体配置」页面的「测试连接」按钮。
    在临时沙箱内启动 CLI + ACP 握手,验证凭证有效性。
    """

    ok: bool
    message: str

File: frontend/src/types/agent_configs.ts
/**
 * 智能体 CLI 配置相关类型定义
 *
 * 对应后端 /agents 系列 API。后端采用动态多 agent 架构:
 * - /agents/types 返回所有已注册 agent 类型及其凭据字段定义(动态表单)
 * - /agents/configs 管理当前用户已配置的 agent(每个 agent_type 一条记录)
 *
 * 当前注册 qoder_cli,未来可扩展 aider / goose 等。
 */

/**
 * 凭据字段定义(后端动态返回,前端据此渲染表单)
 *
 * - secret 类型:password 输入框 + 眼睛切换显隐,加密存储且不回传明文
 * - text 类型:普通输入框,明文存储
 */
export interface CredentialField {
  /** 字段 key,如 'pat' */
  key: string
  /** 展示标签,如 'Personal Access Token' */
  label: string
  /** 输入类型:secret=密码输入+加密存储;text=明文 */
  type: 'secret' | 'text'
  /** 是否必填 */
  required: boolean
  /** 占位提示 */
  placeholder: string
  /** 帮助链接(可选) */
  help_url: string | null
  /** 帮助文本(可选) */
  help_text: string | null
}

/** agent 类型元数据(GET /agents/types) */
export interface AgentTypeMeta {
  /** agent 类型标识,如 'qoder_cli' */
  agent_type: string
  /** 展示名称,如 'Qoder CLI' */
  display_name: string
  /** 描述说明 */
  description: string
  /** 凭据字段定义(动态表单) */
  credential_fields: CredentialField[]
  /** 帮助链接(可选) */
  help_url: string | null
}

/** agent 配置列表项(GET /agents/configs) */
export interface AgentConfigOut {
  /** agent 类型标识 */
  agent_type: string
  /** 展示名称 */
  display_name: string
  /** 是否启用 */
  is_active: boolean
  /** 是否已填写凭据 */
  has_credentials: boolean
}

/** agent 配置详情(GET /agents/configs/{agent_type}、PUT 响应) */
export interface AgentConfigDetailOut extends AgentConfigOut {
  /**
   * 各凭据字段的填写状态
   *
   * key 为字段名,value 为是否已设置。如 { "pat": true }
   */
  credential_status: Record<string, boolean>
}

/** 单个凭据值(保存时提交) */
export interface CredentialValue {
  /** 字段 key */
  key: string
  /**
   * 字段值
   *
   * secret 字段:传空串=保留已存值,传非空=更新;
   * text 字段:直接存。
   */
  value: string
}

/** 保存 agent 配置请求(PUT /agents/configs/{agent_type} body) */
export interface SaveAgentConfigRequest {
  /** 凭据列表 */
  credentials: CredentialValue[]
  /** 是否启用 */
  is_active: boolean
}

/** agent 配置列表响应(GET /agents/configs、DELETE 后响应) */
export interface AgentConfigListResponse {
  configs: AgentConfigOut[]
}

/** 测试 agent 凭证连通性的响应(POST /agents/configs/{agent_type}/test) */
export interface AgentTestResult {
  /** 是否通过 */
  ok: boolean
  /** 人类可读的结果说明(成功/失败原因) */
  message: string
}
